### PR TITLE
 Implement test script

### PR DESCRIPTION
**Pull request by KhanhImma1**
Implement test script for POST_campaign, PUT_channel, GET_all_pixels, helpers files and config files

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 node_modules
 package-lock.json
+.env
```
Comment (evizi-baolonglehoang) [inline .gitignore]: missing endline bro

File: _helpers_/api.helper.ts [hunk 2]
```diff
@@ -13,12 +13,79 @@ export function createCampaign(name: string, slug: string, pub: boolean) {
...
+export async function deletePixel(id: number) {
+    const response = await request(`${base_url}`)
+        .delete("/api/pixel/" + id + "/delete")
+        .set('Authorization', `${token}`);
+    if (response.body.error === 0) {
+        console.log('successfull delete pixel id ' + id);
+    } else {
+        console.log('failed delete pixel id ' + id);
     }
 }
```
Comment (evizi-baolonglehoang) [inline _helpers_/api.helper.ts]: missing endline bro

File: _tests_/GET_all_pixels.spec.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            var randomType = generateRandomType();
+            var randomName = generateRandomString();
+            var randomTag = generateRandomString();
+            const newPixel = await createPixel(randomType, randomName, randomTag);
+            console.log(newPixel.body);
+            pixel_id.push(newPixel.body.id);
+        };
+        const response = await listPixels();
+        expect(response.status).toEqual(200);
+        expect(response.body.error).toEqual(0);
```
Comment (evizi-baolonglehoang) [inline _tests_/GET_all_pixels.spec.ts]: Should create a variable for `response.body`

File: _tests_/GET_all_pixels.spec.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    var pixel_id: number[] = []
+
+    test('PIX-001 Verify that user can get list of total pixels', async function () {
+        const record_size = 2;
+        for (let i = 0; i < record_size; i++) {
+            var randomType = generateRandomType();
+            var randomName = generateRandomString();
+            var randomTag = generateRandomString();
+            const newPixel = await createPixel(randomType, randomName, randomTag);
+            console.log(newPixel.body);
```
Comment (evizi-baolonglehoang) [inline _tests_/GET_all_pixels.spec.ts]: Pls revmove console.log

File: _tests_/GET_all_pixels.spec.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    test('PIX-002 Verify that user can get list of pixels with "limit" in param', async function () {
+        const record_size = 4;
+        const limits = 3;
+        for (let i = 0; i < record_size; i++) {
+            var randomType = generateRandomType();
+            var randomName = generateRandomString();
+            var randomTag = generateRandomString();
+            const newPixel = await createPixel(randomType, randomName, randomTag);
+            console.log(newPixel.body);
```
Comment (evizi-baolonglehoang) [inline _tests_/GET_all_pixels.spec.ts]: same as above

File: _helpers_/common.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+export function generateRandomString() {
+    return Math.random().toString(36).substring(3, 10);
+}
+
+export function generateRandomType() {
```
Comment (kahnext) [inline _helpers_/common.ts]: naming: What type? Pixel Type or something else should be better.

File: _tests_/PUT_channel.spec.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        expect(updatedResponse.status).toEqual(200);
+        expect(updatedResponse.body.error).toEqual(0);
+        expect(updatedResponse.body.name.toLowerCase()).toEqual(randomName_2);
+        expect(updatedResponse.body.description).toEqual(randomDescription_2);
+        expect(updatedResponse.body.color).toEqual("#FFFFFF");
+        expect(updatedResponse.body.starred).toEqual(true);
+        expect(updatedResponse.body).toHaveProperty('id');
+    });
+
+    test('CHA-012 Verify that user cannot change details of an existent channel with existent "name" in Body', async function () {
```
Comment (kahnext) [inline _tests_/PUT_channel.spec.ts]: Not clear test summary. It should Can not rename channel to existing name.

File: _tests_/GET_all_pixels.spec.ts (hunk 4 above)
Comment (KhanhImma1) [reply _tests_/GET_all_pixels.spec.ts]: i would like to check if it's successful to create new records

File: _tests_/PUT_channel.spec.ts (hunk 7 above)
Comment (KhanhImma1) [reply _tests_/PUT_channel.spec.ts]: fixed

File: _helpers_/common.ts (hunk 6 above)
Comment (KhanhImma1) [reply _helpers_/common.ts]: fixed

File: _tests_/GET_all_pixels.spec.ts (hunk 3 above)
Comment (KhanhImma1) [reply _tests_/GET_all_pixels.spec.ts]: fixed

File: _helpers_/api.helper.ts (hunk 2 above)
Comment (KhanhImma1) [reply _helpers_/api.helper.ts]: fixed

File: .gitignore (hunk 1 above)
Comment (KhanhImma1) [reply .gitignore]: fixed

File: _helpers_/const.ts [hunk 8]
```diff
@@ -1,8 +1,6 @@
...
-export var randomSlug = Math.random().toString(36).substring(2, 7);
-export var randomDescription = Math.random().toString(36).substring(4, 14);
-const types = ["gapixel", "adwordspixel", "twitterpixel", "adrollpixel", "pinterest", "bing", "snapchat", "reddit", "tiktok"];
-export var randomType = types[Math.floor(Math.random() * types.length)];
-export var randomTag = Math.random().toString(36).substring(2, 7);
\ No newline at end of file
+import * as dotenv from 'dotenv'
+dotenv.config();
+
+export const base_url = process.env.BASE_URL;
```
Comment (trungtranevizi) [inline _helpers_/const.ts]: Constant variable should be in uppercase

File: _helpers_/const.ts [hunk 9]
```diff
@@ -1,8 +1,6 @@
...
+export const token = process.env.TOKEN;
```
Comment (trungtranevizi) [inline _helpers_/const.ts]: same above

File: _tests_/GET_all_pixels.spec.ts (hunk 4 above)
Comment (trungtranevizi) [reply _tests_/GET_all_pixels.spec.ts]: you use console.log for debugging purpose. You shouldn't push this.

Comment (KhanhImma1) [reply _tests_/GET_all_pixels.spec.ts]: i got it

File: _helpers_/const.ts (hunk 9 above)
Comment (KhanhImma1) [reply _helpers_/const.ts]: fixed

File: _helpers_/const.ts (hunk 8 above)
Comment (KhanhImma1) [reply _helpers_/const.ts]: fixed

File: _tests_/GET_all_pixels.spec.ts (hunk 5 above)
Comment (KhanhImma1) [reply _tests_/GET_all_pixels.spec.ts]: fixed

**Review by kahnext — APPROVED** [review]
LGTM

File: helpers/api.helper.ts [hunk 10]
```diff
@@ -47,14 +45,12 @@ export function updateChannel(id: number, name: string, des: string, color: stri
...
-export async function deleteChannel(id: number) {
-    const response = await request(BASE_URL)
-        .delete("/api/channel/" + id + "/delete")
-        .set('Authorization', `${TOKEN}`);
-    if (response.body.error === 0) {
-        console.log('successfull delete channel id ' + id);
-    } else {
-        console.log('failed delete channel id ' + id);
+export async function deleteManyChannels(array: string[]) {
+    const length = array.length;
```
Comment (thanhevizi) [inline helpers/api.helper.ts]: No needs to create the "length" variable here, we only use it one time for this function

File: __tests__/get_all_pixels.spec.ts [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        const response = await request(BASE_URL)
+            .get("/api/pixels")
+            .set('Authorization', `${TOKEN}`);
+        const data = response.body.data;
+        expect(response.status).toEqual(200);
+        expect(response.body.error).toEqual(0);
+        expect(data.result).toEqual(record_size);
+        expect(data.perpage).toEqual(15);
+        expect(data.currentpage).toEqual(1);
+        expect(data.nextpage).toEqual(null);
```
Comment (thanhevizi) [inline __tests__/get_all_pixels.spec.ts]: No needs to verify value of "record_size", "nextpage" and "maxpage". This should be verify a different test cases. For scope of this test case, just verifying body contains these fields.

File: __tests__/get_all_pixels.spec.ts [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+                limit: limits
+            });
+        const data = response.body.data;
+        expect(response.status).toEqual(200);
+        expect(response.body.error).toEqual(0);
+        expect(data.result).toEqual(record_size);
+        expect(data.perpage).toEqual(limits);
+        expect(data.currentpage).toEqual(1);
+        expect(data.nextpage).toEqual(2);
+        expect(data.maxpage).toEqual(2);
```
Comment (thanhevizi) [inline __tests__/get_all_pixels.spec.ts]: Same as above test case

File: __tests__/get_all_pixels.spec.ts [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+import { createPixel, deleteManyPixels } from '../helpers/api.helper'
+import { DEFAULT_TIMEOUT } from '../helpers/constant'
+
+describe('GET /api/pixels', function () {
+    var pixel_id: string[] = []
+
+    test('PIX-001 Verify that user can get list of total pixels', async function () {
+        const record_size = 2;
+        // create new pixels for testing
+        for (let i = 0; i < record_size; i++) {
```
Comment (thanhevizi) [inline __tests__/get_all_pixels.spec.ts]: This step should be in beforeAll or beforeEach as pre-condition steps

File: __tests__/get_all_pixels.spec.ts [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+        expect(typeof data.pixels[0].date).toBe('string');
+    },
+        DEFAULT_TIMEOUT
+    )
+
+    test('PIX-002 Verify that user can get list of pixels with "limit" in param', async function () {
+        const record_size = 4;
+        const limits = 3;
+        // create new pixels for testing
+        for (let i = 0; i < record_size; i++) {
```
Comment (thanhevizi) [inline __tests__/get_all_pixels.spec.ts]: Same as above test case

File: __tests__/put_channel.spec.ts [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+describe('PUT /api/channel/:id/update', function () {
+    var channel_id: string[] = [];
+
+    test('CHA-001 Verify that user can change total details of an existent channel', async function () {
+        const randomName = generateRandomString();
+        const randomDescription = generateRandomString();
+        const randomName_2 = generateRandomString();
+        const randomDescription_2 = generateRandomString();
+        // create a new channel for testing
+        const createdResponse = await createChannel(randomName, randomDescription, "#000000", false);
```
Comment (thanhevizi) [inline __tests__/put_channel.spec.ts]: This step should be in beforeEach or beforeAll as pre-condition steps

File: __tests__/put_channel.spec.ts [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+
+    test('CHA-012 Verify that user cannot rename an existent channel with existent "name" in Body', async function () {
+        const randomName = generateRandomString();
+        const randomDescription = generateRandomString();
+        const randomName_2 = generateRandomString();
+        const randomDescription_2 = generateRandomString();
+        // created 2 new channels
+        const newChannel_1 = await createChannel(randomName, randomDescription, "#000000", true);
+        const channel_id_1 = newChannel_1.body.id;
+        const newChannel_2 = await createChannel(randomName_2, randomDescription_2, "#000000", false);
```
Comment (thanhevizi) [inline __tests__/put_channel.spec.ts]: These steps should be in beforeEach or beforeAll as pre-condition steps

File: helpers/api.helper.ts (hunk 10 above)
Comment (KhanhImma1) [reply helpers/api.helper.ts]: fixed

File: __tests__/get_all_pixels.spec.ts (hunk 11 above)
Comment (KhanhImma1) [reply __tests__/get_all_pixels.spec.ts]: fixed

File: __tests__/get_all_pixels.spec.ts (hunk 12 above)
Comment (KhanhImma1) [reply __tests__/get_all_pixels.spec.ts]: fixed

File: __tests__/get_all_pixels.spec.ts (hunk 13 above)
Comment (KhanhImma1) [reply __tests__/get_all_pixels.spec.ts]: fixed

File: __tests__/get_all_pixels.spec.ts (hunk 14 above)
Comment (KhanhImma1) [reply __tests__/get_all_pixels.spec.ts]: fixed

File: __tests__/put_channel.spec.ts (hunk 15 above)
Comment (KhanhImma1) [reply __tests__/put_channel.spec.ts]: fixed

File: __tests__/put_channel.spec.ts (hunk 16 above)
Comment (KhanhImma1) [reply __tests__/put_channel.spec.ts]: fixed

**Pull request merged**